topics/authentication.md

```diff
@@ -6,140 +6,151 @@ Code examples: <a href="samples.md" anchor="authentication">Authentication</a>
+
 
+### Step 2: Specify a provider name {id="provider-name"}
 
-After defining one or more authentication providers (named or unnamed), with the [routing plugin](Routing_in_Ktor.md)
-you can create a route group, that will apply that authentication to all the routes defined in that group:
+A function for [using a specific provider](#choose-provider) optionally allows you to specify a provider name. A code sample below installs the [basic](https://api.ktor.io/ktor-features/ktor-auth/ktor-auth/io.ktor.auth/basic.html) and [form](https://api.ktor.io/ktor-features/ktor-auth/ktor-auth/io.ktor.auth/form.html) providers with the _auth-basic_ and _auth-form_ names, respectively:
 
 ```kotlin
-routing {
-    authenticate("myauth1") {
-        get("/authenticated/route1") {
-            // ...
-        }    
-        get("/other/route2") {
-            // ...
-        }    
+install(Authentication) {
+    basic("auth-basic") {
+        // [[[Configure basic authentication|basic.md]]]
     }
-    get("/") {
-        // ...
+    form("auth-form") {
+        // [[[Configure form authentication|form.md]]]
     }
+    // ...
 }
 ```
+{disable-links="false"}
 
-You can specify several names to apply several authentication providers, or none or null to use the unnamed one.
+These names can be used later to [authenticate different routes](#authenticate-route) using different providers.
+> Note that a provider name should be unique, and you can define only one provider without a name.
 
-You can get the generated `Principal` instance inside your handler with:
 
-```kotlin
-val principal: UserIdPrincipal? = call.authentication.principal<UserIdPrincipal>()
-```
+### Step 3: Configure a provider {id="configure-provider"}
 
->In the generic, you have to put a specific type that *must* match the generated Principal.
->It will return null in the case you provide another type.
->
-{type="note"}
+Each [provider type](#choose-provider) has its own configuration. For instance, the [BasicAuthenticationProvider.Configuration](https://api.ktor.io/ktor-features/ktor-auth/ktor-auth/io.ktor.auth/-basic-authentication-provider/-configuration/index.html) class contains options passed to the [basic](https://api.ktor.io/ktor-features/ktor-auth/ktor-auth/io.ktor.auth/basic.html) function. The most important function exposed by this class is [validate](https://api.ktor.io/ktor-features/ktor-auth/ktor-auth/io.ktor.auth/-basic-authentication-provider/-configuration/validate.html) that validates a user name and password. A code sample below shows how it can look:
 
->The handler won't be executed if the configured authentication fails (when returning `null` in the authentication mechanism)
->
-{type="note"}
+```kotlin
+```
+{src="snippets/auth-basic/src/main/kotlin/com/example/Application.kt" lines="9-20"}
 
-## Naming the authentication provider
+To understand how the `validate` function works, we need to introduce two terms:
```

Comment:
This feels strange, because it feels `principal` is part of Basic authentication and it's not. I would clear up the fact that the above is an example of Basic authentication, and that principal is a generic concept.